### PR TITLE
[GOBBLIN-1377] Adds functionality to create shards for target directories in hive distcp

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1377

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):

This PR allows Gobblin to have an entrypoint to create shards on cloud environments before copying. Notable changes include
1) A configuration `gobblin.useDatasetLocalWorkDir` to enable dataset-local temporary directories for the writer and error files instead of a storing these on the job level.
2) Refactoring of `HiveCopyEntityHelper` for code cleanup
3) Hive Dataset unit tests for `CopySource`

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 1]
```diff
@@ -351,11 +358,21 @@
         } else {
           this.existingTargetTable = Optional.absent();
         }
-
```
Comment (aplex) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java:354]: This constructor is too complex, and it's hard to follow what's going on here. You can look if it can be refactored and split into smaller functions. If the refactoring is big and touches a lot of code, you can send it as a separate PR, before you main change.

You can also install this IntelliJ plugin that will tell you if the function is too complex, right in the IDE: https://plugins.jetbrains.com/plugin/12159-codemetrics

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveTargetDirectoryClient.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package org.apache.gobblin.data.management.copy.hive;
+
+import java.io.IOException;
+import java.util.Properties;
+import org.apache.hadoop.fs.Path;
+
+
+public class HiveTargetDirectoryClient {
```
Comment (aplex) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveTargetDirectoryClient.java:8]: You can add some docs all around, explaining how this can be used and extended

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveTargetDirectoryClient.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  public Path getOrCreateTargetPath(Path path) throws IOException {
+    return path;
+  }
+
+  /**
+   * Deletes references to the path from the directory client
+   * @param path path on target directory
+   * @throws IOException if deleting the path fails
+   */
+  public void deletePath(Path path) throws IOException {
```
Comment (aplex) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveTargetDirectoryClient.java:32]: This is a surprising behavior - we call delete and nothing happens by default. I would expect it to delete a thing.

Let's think about how this interface can be made more obvious.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveTargetDirectoryClient.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.Properties;
+import org.apache.hadoop.fs.Path;
+
+
+/**
+ * Determines and manages target paths for {@link HiveCopyEntityHelper}
+ * An example implementation would create target directories on cloud environments before initiating the copy
+ * By default, returns the input path (do nothing)
+ */
+public class HiveTargetDirectoryClient {
```
Comment (aplex) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveTargetDirectoryClient.java:13]: This can be an interface and a default implementation class.

**Review by aplex — COMMENTED** [review]
Things look better after refactoring.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java [hunk 5]
```diff
@@ -197,6 +197,14 @@ public static void cleanTaskStagingData(State state, Logger logger) throws IOExc
           throw new IOException("Clean up output directory " + outputPath.toUri().getPath() + " failed");
         }
       }
+
+      Path errPath = new Path(state.getProp(ConfigurationKeys.ROW_LEVEL_ERR_FILE));
```
Comment (aplex) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java:201]: Is this cleanup related to sharding, or it is some separate improvement?

Also, is this error file not part of staging/output folder that are cleaned up earlier in this function?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java [hunk 6]
```diff
@@ -370,9 +366,25 @@ public Void call() {
               workUnit.setProp(ConfigurationKeys.COPY_EXPECTED_SCHEMA, ((ConfigBasedDataset) this.copyableDataset).getExpectedSchema());
             }
           }
-          if ((this.copyableDataset instanceof HiveDataset) && (state.getPropAsBoolean(ConfigurationKeys.IS_DATASET_STAGING_DIR_USED,false))) {
-            workUnit.setProp(DATASET_STAGING_DIR_PATH, ((HiveDataset) this.copyableDataset).getProperties().getProperty(DATASET_STAGING_PATH));
+
+          // Ensure that the writer temporary directories are contained within the dataset shard
+          if ((this.copyableDataset instanceof HiveDataset) && (state.getPropAsBoolean(ConfigurationKeys.USE_SHARDED_WRITER_DIRS,false))) {
```
Comment (aplex) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java:371]: Would be good to add a bunch of unit tests, verifying that dataset-local-staging flag works as expect. This should help us iron out typos before doing end-to-end tests.

File: gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java [hunk 7]
```diff
@@ -210,7 +210,6 @@
   public static final String MAXIMUM_JAR_COPY_RETRY_TIMES_KEY = JOB_JAR_FILES_KEY + ".uploading.retry.maximum";
   public static final String USER_DEFINED_STATIC_STAGING_DIR = "user.defined.static.staging.dir";
   public static final String USER_DEFINED_STAGING_DIR_FLAG = "user.defined.staging.dir.flag";
-  public static final String IS_DATASET_STAGING_DIR_USED = "dataset.staging.dir.used";
```
Comment (aplex) [inline gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java:213]: Note: need to update the docs/wiki pages after this is merged, to reflect the new name.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java [hunk 8]
```diff
@@ -370,9 +368,16 @@ public Void call() {
...
+            String datasetPath = ((HiveDataset) this.copyableDataset).getProperties().getProperty(DATASET_STAGING_PATH);
+            workUnit.setProp(ConfigurationKeys.WRITER_STAGING_DIR, datasetPath + ConfigurationKeys.STAGING_DIR_DEFAULT_SUFFIX + "/" + state
```
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java]: @aplex Yes I was thinking whether or not to allow both configs and have the `WRITER_STAGING_DIR` config work as specifying the suffix when `USE_SHARDED_WRITER_DIRS` is enabled, or to keep them fully mutually exclusive. Will ask SRE for their input on this as it'll cause quite a few configurations to be modified, but that's be unavoidable regardless what we go with.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveDataset.java [hunk 9]
```diff
@@ -130,7 +130,9 @@ public HiveDataset(FileSystem fs, HiveMetastoreClientPool clientPool, Table tabl
     this.tableIdentifier = this.table.getDbName() + "." + this.table.getTableName();
     Path tableLocation = this.table.getPath();
     if (!(this.properties.isEmpty())) {
```
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveDataset.java:132]: Yes you're right I'll check specifically for the added flag in the state instead, not sure why it was implemented like this @autumnust if you have more context.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java (hunk 5 above)
Comment (Will-Lo) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java:201]: The error file is specified as a separate path from the writer output/staging dirs, so I considered it as another temporary job-level directory that has to be modified to work on the dataset-level. Hence why it has to be added in cleanup here.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java (hunk 6 above)
Comment (Will-Lo) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java:371]: Yes I'll add the unit tests to the source file. Wasn't sure if it was going to be done on the source side or writer side but the tests would be similar regardless where we implement the logic in the end.

File: gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java [hunk 10]
```diff
@@ -1062,4 +1061,11 @@
    */
   public static final String TASK_EVENT_METADATA_GENERATOR_CLASS_KEY = "gobblin.task.event.metadata.generator.class";
   public static final String DEFAULT_TASK_EVENT_METADATA_GENERATOR_CLASS_KEY = "nooptask";
+
+  /**
+   * Configuration for sharded directory files
+   */
+  public static final String USE_SHARDED_WRITER_DIRS = "gobblin.dataset.sharded.work.dirs";
```
[the remark this answers is not in the document]
Comment (Will-Lo) [reply gobblin-api/src/main/java/org/apache/gobblin/configuration/ConfigurationKeys.java:1068]: I like `gobblin.useDatasetLocalWorkDir`, thanks for the suggestion :)

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 11]
```diff
@@ -225,8 +226,13 @@ private void publishFileSet(CopyEntity.DatasetAndPartition datasetAndPartition,
...
-    CopyableDatasetMetadata metadata = CopyableDatasetMetadata
-        .deserialize(datasetWorkUnitStates.iterator().next().getProp(CopySource.SERIALIZED_COPYABLE_DATASET));
+    CopyableDatasetMetadata metadata = CopyableDatasetMetadata.deserialize(
+        datasetWorkUnitStates.iterator().next().getProp(CopySource.SERIALIZED_COPYABLE_DATASET));
+
+    // Get the dataset specific sharded directory if configured
+    if (state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR)) {
+      this.writerOutputDir = new Path(datasetWorkUnitStates.iterator().next().getProp(ConfigurationKeys.WRITER_OUTPUT_DIR));
+    }
     Path datasetWriterOutputPath = new Path(this.writerOutputDir, datasetAndPartition.identifier());
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java:236]: 1. Can we store `datasetWorkUnitStates.iterator().next()` in a temp variable like `sampledWUS` to avoid calling twice? 
2. `writerOutputDir` in this case should be a local variable since it will be specific to a workunitstate. Maybe converting the whole thing including line 236 into a ternary operator like `Path datasetWriterOutputPath = state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR) ? xxx : this.writerOutputDir` ? just leave the `writerOutputDir` as null otherwise might be safer: we can risk fail the job but not writing something that are not supposed to be in a dataset folder

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 12]
```diff
@@ -79,7 +79,7 @@
 @Slf4j
 public class CopyDataPublisher extends DataPublisher implements UnpublishedHandling {
 
-  private final Path writerOutputDir;
+  private Path writerOutputDir;
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java:82]: if adopting suggestion below this can stay as final I guess?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java [hunk 13]
```diff
@@ -370,9 +366,25 @@ public Void call() {
               workUnit.setProp(ConfigurationKeys.COPY_EXPECTED_SCHEMA, ((ConfigBasedDataset) this.copyableDataset).getExpectedSchema());
             }
           }
-          if ((this.copyableDataset instanceof HiveDataset) && (state.getPropAsBoolean(ConfigurationKeys.IS_DATASET_STAGING_DIR_USED,false))) {
-            workUnit.setProp(DATASET_STAGING_DIR_PATH, ((HiveDataset) this.copyableDataset).getProperties().getProperty(DATASET_STAGING_PATH));
+
+          // Ensure that the writer temporary directories are contained within the dataset shard
+          if ((this.copyableDataset instanceof HiveDataset) && (state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR,false))) {
```
Comment (autumnust) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java:371]: It is OK to do it temporarily but using `instanceOf` in `CopySource` definitely seems like a abstraction leaking ( I know it was there already). 

We don't have to address this in this PR, but do you think we should put something like `getDatasetPath` as part of `CopyableDatasetBase` interface?  So that we don't need to explicitly cast them

**Review by autumnust — CHANGES_REQUESTED** [review]
Haven't gone through it fully. This is my first scan.

**Comment by sv2000** [comment]
@Will-Lo Thanks for the PR. Going through the PR, I have a different proposal for how dataset-specific logic can be injected. We could invoke one or more configurable handler classes inside the AbstractJobLauncher right after the workunit creation step. For your use case, you could use the DATASET_URN property which is set in different source classes. You could provide an implementation of a handler that sets dataset-specific staging/output dirs. Please take a look at the TaskStateCollectorServiceHandler class as an example, which is invoked on task completion. What I am proposing here is an identical solution, but invoked before a job is launched.  

This high level approach is leverageable for future use cases e.g. where a Gobblin pipeline needs to write to a Table (e.g. Iceberg) and we could define a handler that ensures table is created before the job is started. 

Happy to discuss offline if you have questions.

File: gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandlerService.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+ * Reads {@link ConfigurationKeys#DESTINATION_DATASET_HANDLER_CLASS} as a list
+ * of classes, separated by comma to initialize the handlers
+ */
+public class DestinationDatasetHandlerService implements Closeable {
+  List<DestinationDatasetHandler> handlers;
+
+  public DestinationDatasetHandlerService(SourceState jobState) {
+    this.handlers = new ArrayList<>();
+    if (jobState.contains(ConfigurationKeys.DESTINATION_DATASET_HANDLER_CLASS)) {
+      String[] handlerList = jobState.getProp(ConfigurationKeys.DESTINATION_DATASET_HANDLER_CLASS).split(",");
```
Comment (aplex) [inline gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandlerService.java:42]: I suggest to trim the whitespaces from splitted values . I've seen that people like to separate multiple value in configs with a coma and a space -"value1, value2, value3".

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java [hunk 15]
```diff
@@ -370,9 +366,13 @@ public Void call() {
               workUnit.setProp(ConfigurationKeys.COPY_EXPECTED_SCHEMA, ((ConfigBasedDataset) this.copyableDataset).getExpectedSchema());
             }
           }
-          if ((this.copyableDataset instanceof HiveDataset) && (state.getPropAsBoolean(ConfigurationKeys.IS_DATASET_STAGING_DIR_USED,false))) {
-            workUnit.setProp(DATASET_STAGING_DIR_PATH, ((HiveDataset) this.copyableDataset).getProperties().getProperty(DATASET_STAGING_PATH));
+
+          // Ensure that the writer temporary directories are contained within the dataset shard
+          if (state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR,false)) {
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java:371]: Should we just set DATASET_DESTINATION_PATH regardless of whether we want to use DATASET_LOCAL_WORK_DIR?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 16]
```diff
@@ -354,35 +313,24 @@
           this.existingTargetTable = Optional.absent();
         }
 
-        // Constructing CommitStep object for table registration
-        Path targetPath = getTargetLocation(this.dataset.fs, this.targetFs, this.dataset.table.getDataLocation(),
-            Optional.<Partition> absent());
+        Path targetPath = getTargetLocation(this.targetFs, this.dataset.table.getDataLocation(), Optional.<Partition>absent());
+        if (Boolean.parseBoolean(this.dataset.getProperties().getProperty(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR, "false"))) {
```
Comment (aplex) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java:317]: We have a bunch of utils to work with config/properties/maps - ConfigUtils.getBoolean() , State.getPropAsBoolean()
If none of them work here, you can use  MapUtils.getBooleanValue from Apache Commons.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 17]
```diff
@@ -117,8 +117,9 @@ public CopyDataPublisher(State state) throws IOException {
     this.fs = FileSystem.get(URI.create(uri), WriterUtils.getFsConfiguration(state));
 
     FileAwareInputStreamDataWriterBuilder.setJobSpecificOutputPaths(state);
-
-    this.writerOutputDir = new Path(state.getProp(ConfigurationKeys.WRITER_OUTPUT_DIR));
+    // If directories are sharded by dataset, initialize writers using workunit state instead
+    this.writerOutputDir = state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR) ?
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java:121]: Shouldn't WRITER_OUTPUT_DIR be set inside the handler? If so, we don't need these changes here, do we?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriterBuilder.java [hunk 18]
```diff
@@ -52,14 +53,13 @@
...
     if (!StringUtils.containsIgnoreCase(state.getProp(ConfigurationKeys.WRITER_STAGING_DIR),
-        state.getProp(ConfigurationKeys.JOB_ID_KEY))) {
-
-      state.setProp(ConfigurationKeys.WRITER_STAGING_DIR, new Path(state.getProp(ConfigurationKeys.WRITER_STAGING_DIR),
-          state.getProp(ConfigurationKeys.JOB_ID_KEY)));
-      state.setProp(ConfigurationKeys.WRITER_OUTPUT_DIR, new Path(state.getProp(ConfigurationKeys.WRITER_OUTPUT_DIR),
-          state.getProp(ConfigurationKeys.JOB_ID_KEY)));
-
+        state.getProp(ConfigurationKeys.JOB_ID_KEY)) && !state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR)) {
+        state.setProp(ConfigurationKeys.WRITER_STAGING_DIR, new Path(state.getProp(ConfigurationKeys.WRITER_STAGING_DIR),
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriterBuilder.java:59]: Can all of this be moved inside the handler?

**Review by sv2000 — CHANGES_REQUESTED** [review]
Is looking better. But I think we should aim mostly to leave source/writer/publishers unchanged and move dataset-specific logic inside the DatasetHandler impl.

File: gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandlerService.java (hunk 14 above)
Comment (Will-Lo) [reply gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandlerService.java:42]: This is a really good point, thanks for the catch

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java (hunk 17 above)
Comment (sv2000) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java:121]: I don't understand the explanation. What I am suggesting is to have the ConfigurationKeys#WRITER_OUTPUT_DIR being overwritten inside the DatasetHandler impl, which would avoid the need for this change.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriterBuilder.java (hunk 18 above)
Comment (sv2000) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriterBuilder.java:59]: Same comment as earlier. We should be able to set override the WRITER_STAGING_DIR config inside the DatasetHandler and leave the code here unchanged.

**Review by sv2000 — CHANGES_REQUESTED** [review]

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java [hunk 19]
```diff
@@ -912,13 +915,14 @@ private void cleanLeftoverStagingData(WorkUnitStream workUnits, JobState jobStat
     }
 
     try {
-      if (this.jobContext.shouldCleanupStagingDataPerTask()) {
+      if (this.jobContext.shouldCleanupStagingDataPerTask() || jobState.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR)) {
```
Comment (sv2000) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java:918]: Why do we need to change the condition in the if (..) statement?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java [hunk 20]
```diff
@@ -972,7 +974,7 @@ private void cleanupStagingData(JobState jobState)
       throw new JobException("Failed to check unfinished commit sequences", e);
     }
 
-    if (this.jobContext.shouldCleanupStagingDataPerTask()) {
+    if (this.jobContext.shouldCleanupStagingDataPerTask() || jobState.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR)) {
```
Comment (sv2000) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java:977]: Same comment as above.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java [hunk 21]
```diff
@@ -197,6 +197,16 @@ public static void cleanTaskStagingData(State state, Logger logger) throws IOExc
           throw new IOException("Clean up output directory " + outputPath.toUri().getPath() + " failed");
         }
       }
+
+      if (state.contains(ConfigurationKeys.ROW_LEVEL_ERR_FILE) && state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR, false)) {
```
Comment (sv2000) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java:201]: We should avoid leaking details about dataset-specific local work dir all over the place. This change seems unnecessary too.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java [hunk 22]
```diff
@@ -235,6 +245,14 @@ public static void cleanTaskStagingData(State state, Logger logger, Closer close
         logger.info("Cleaning up output directory " + outputPath.toUri().getPath());
         parallelRunner.deletePath(outputPath, true);
       }
+
+      if (state.contains(ConfigurationKeys.ROW_LEVEL_ERR_FILE) && state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR, false)) {
```
Comment (sv2000) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java:249]: same comment as earlier.

**Review by sv2000 — CHANGES_REQUESTED** [review]

File: gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandler.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.Collection;
+import org.apache.gobblin.source.workunit.WorkUnit;
+
+/**
+ * Performs work related to initializing the target environment before the files are written and published
+ */
+public interface DestinationDatasetHandler extends Closeable {
+
+  /**
+   * Adds required properties to work units for determining writer staging and output paths
```
Comment (autumnust) [inline gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandler.java:31]: The purpose for this class doesn't have to be very specific to the sharding thing.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriterBuilder.java (hunk 18 above)
Comment (autumnust) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriterBuilder.java:59]: @Will-Lo  If your concern is about hive-distcp user with shard configuration: It should be fine to move this into handler only since I don't think there's any real usage existed for shard-aware distcp

**Review by autumnust — CHANGES_REQUESTED** [review]

File: gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandlerService.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+ * Reads {@link ConfigurationKeys#DESTINATION_DATASET_HANDLER_CLASS} as a list
+ * of classes, separated by comma to initialize the handlers
+ */
+public class DestinationDatasetHandlerService implements Closeable {
+  List<DestinationDatasetHandler> handlers;
+
+  public DestinationDatasetHandlerService(SourceState jobState, Boolean canCleanUp, EventSubmitter eventSubmitter) {
+    this.handlers = new ArrayList<>();
+    if (jobState.contains(ConfigurationKeys.DESTINATION_DATASET_HANDLER_CLASS)) {
+      String[] handlerList = jobState.getProp(ConfigurationKeys.DESTINATION_DATASET_HANDLER_CLASS)
```
Comment (sv2000) [inline gobblin-core/src/main/java/org/apache/gobblin/destination/DestinationDatasetHandlerService.java:44]: PropertiesUtils.getPropAsList()?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopyableDatasetBase.java [hunk 25]
```diff
@@ -18,11 +18,13 @@
 package org.apache.gobblin.data.management.copy;
 
 import org.apache.gobblin.dataset.Dataset;
+import org.apache.hadoop.fs.Path;
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopyableDatasetBase.java:21]: Unused import?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java [hunk 26]
```diff
@@ -354,35 +314,24 @@
           this.existingTargetTable = Optional.absent();
         }
 
-        // Constructing CommitStep object for table registration
-        Path targetPath = getTargetLocation(this.dataset.fs, this.targetFs, this.dataset.table.getDataLocation(),
-            Optional.<Partition> absent());
+        Path targetPath = getTargetLocation(this.targetFs, this.dataset.table.getDataLocation(), Optional.<Partition>absent());
+        if (MapUtils.getBoolean(this.dataset.getProperties(), ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR, false)) {
+          this.dataset.datasetPath = targetPath.toUri().getRawPath();
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java:319]: Why does datasetPath depend on whether or not we are using dataset local working dir?

File: gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java [hunk 27]
```diff
@@ -132,10 +132,9 @@ public static void cleanStagingData(List<? extends State> states, Logger logger)
    * @param logger a {@link Logger} used for logging
    */
   public static void cleanJobStagingData(State state, Logger logger) throws IOException {
-    Preconditions.checkArgument(state.contains(ConfigurationKeys.WRITER_STAGING_DIR),
-        "Missing required property " + ConfigurationKeys.WRITER_STAGING_DIR);
-    Preconditions.checkArgument(state.contains(ConfigurationKeys.WRITER_OUTPUT_DIR),
-        "Missing required property " + ConfigurationKeys.WRITER_OUTPUT_DIR);
+    if (!state.contains(ConfigurationKeys.WRITER_STAGING_DIR) || !state.contains(ConfigurationKeys.WRITER_OUTPUT_DIR)) {
```
Comment (sv2000) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java:135]: Shouldn't this be invoked after the workunit stream has been computed? If so, shouldn't we expect both these config keys to be set?

Comment (sv2000) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java:135]: This looks like a change in behavior from what existed earlier.

**Review by sv2000 — CHANGES_REQUESTED** [review]

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveDataset.java [hunk 28]
```diff
@@ -86,14 +86,15 @@
...
   public static final String TABLE = "Table";
-  public static final String DATASET_STAGING_PATH = "dataset.staging.path";
 
   public static final String DATABASE_TOKEN = "$DB";
   public static final String TABLE_TOKEN = "$TABLE";
 
   public static final String LOGICAL_DB_TOKEN = "$LOGICAL_DB";
   public static final String LOGICAL_TABLE_TOKEN = "$LOGICAL_TABLE";
 
+  public String datasetPath;
```
Comment (sv2000) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/hive/HiveDataset.java:96]: Should this be private?

**Review by sv2000 — CHANGES_REQUESTED** [review]

File: gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java (hunk 27 above)
Comment (Will-Lo) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/JobLauncherUtils.java:135]: This checks the job level state, not the workunit. So it would fail if we do not specify the temporary directories.

**Review by sv2000 — APPROVED** [review]
+1. LGTM.

**Review by autumnust — APPROVED** [review]
